state/sqlserver/sqlserver.go

```diff
@@ -453,8 +453,8 @@ func (s *SQLServer) Delete(req *state.DeleteRequest) error {
 		return err
 	}
 
-	if rows != 1 {
-		return fmt.Errorf("items was not updated")
+	if req.ETag != nil && rows != 1 {
+		return state.NewETagError(state.ETagMismatch, nil)
```

Comment:
The existing error check is still valid independent of ETags, so it may be useful to preserve the separation of errors. There is also a question of whether empty string ETags are considered mismatches in this case, since elsewhere in the code, it always collapses empty strings into the nil case.

```suggestion
	rows != 1 {
		if req.ETag != nil && *req.ETag != "" {
			return state.NewETagError(state.ETagMismatch, nil)
		} else {
			return fmt.Errorf("items was not updated")
		}
	}
```

Reply:
I agree, if an updated was expected without eTag but nothing happened, it is also an error.

Reply:
I specifically asked you in a meeting about this behavior @artursouza - whether a deletion of a record which doesn't exist can be a no-op.

The reason being that the conformance tests attempt to delete a bunch of records that were not actually inserted but expect those deletions to succeed (no op I guess)

Reply:
The Delete test depends on the Set test, and the set test inserts a different set then is being deleted in the delete test.

https://github.com/dapr/components-contrib/blob/master/tests/conformance/state/state.go#L207

https://github.com/dapr/components-contrib/blob/master/tests/conformance/state/state.go#L237

Clearly the other components are not throwing an error when a record is deleted which does not exist!

Reply:
I see two choices -- either those conformance tests need to be fixed to expect failures of deletions which do not exist (I would prefer this), or this integration needs to be updated not to expect a failure when a deletion is made with an invalid ETAG: https://github.com/dapr/components-contrib/blob/master/state/sqlserver/sqlserver_integration_test.go#L245

Reply:
Any preferences here @artursouza @CodeMonkeyLeet ?

I wonder what the fallout is if we correct the conformance test.

Reply:
Changing the test suggests a breaking change in the error contract with the app (conformance tests are supposed to check adherence to it after all).

The problem in the existing SQL server implementation specifically is that it doesn't differentiate between the delete failing because:
- the item did not exist (no-op) and 
- the item exists but delete targeted the wrong version (dapr wants to error on the ETag mismatch).

If the code only checks the presence of ETag on failed delete, it could be a false negative (failure could be because the item did not exist, ETag or otherwise). Instead of checking for the deleted rows, is there a try-catch mechanism available to determine the cause of failure?

Reply:
Thanks for taking the time to have a call with me @CodeMonkeyLeet.

I'll summarize the conversation / these tests again here:

The first think we do is throw any actual database level error returned. Now it's important to note that SQLServer doesn't throw an error when a key to be deleted doesn't exist.

Where the key doesn't exist we receive a success but with rows Affected = 0. The conformance tests expect a Delete Request where an item did not exist to succeed (return nil for the error). So in the case where we don't have an ETag we can say that rows = 0 or rows = 1 are acceptable states and therefore we return `nil`. Due to the primary key constraint we can't possibly have more rows affected, so therefore we can always return `nil` if there was no database error and no ETag provided.

In the case where an ETag is provided the question is whether a deletion which did not actually remove a row (because the key didn't exist) is an error. I believe yes -- and an existing integration test (`state/sqlserver/sqlserver_integration_test.go`) specifically tests for this error. For this particular reason I'm throwing an ETag error in the case where the tag was provided but no record delete (but again no failure from the database itself).

I believe this is the correct behavior - and the conformance tests and the existing integration test would agree.